File: mobile/android/UI/GlassmorphismEffects.kt
/*
 * GlassmorphismEffects.kt
 * ALADDIN Mobile Security
 *
 * Glassmorphism UI Effects for Android
 * Modern glass-like visual effects with blur and transparency
 *
 * Created by ALADDIN Security Team
 * Date: 2025-01-27
 * Version: 1.0
 */

package com.aladdin.security.ui.glassmorphism

import android.content.Context
import android.graphics.*
import android.graphics.drawable.Drawable
import android.graphics.drawable.GradientDrawable
import android.os.Build
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import androidx.core.graphics.drawable.DrawableCompat
import androidx.core.view.ViewCompat
import androidx.core.view.setPadding

// MARK: - Glassmorphism Effect Types
enum class GlassmorphismStyle {
    LIGHT,      // Light glass effect
    MEDIUM,     // Medium glass effect
    HEAVY,      // Heavy glass effect
    FROSTED,    // Frosted glass effect
    CRYSTAL,    // Crystal clear effect
    STORM       // Storm sky themed effect
}

// MARK: - Glassmorphism Configuration
data class GlassmorphismConfig(
    val style: GlassmorphismStyle,
    val blurRadius: Float,
    val opacity: Float,
    val cornerRadius: Float,
    val borderWidth: Float,
    val borderColor: Int,
    val shadowRadius: Float,
    val shadowOpacity: Float,
    val shadowOffsetX: Float,
    val shadowOffsetY: Float,
    val shadowColor: Int
) {
    companion object {
        val LIGHT = GlassmorphismConfig(
            style = GlassmorphismStyle.LIGHT,
            blurRadius = 10f,
            opacity = 0.1f,
            cornerRadius = 12f,
            borderWidth = 1f,
            borderColor = Color.argb(51, 255, 255, 255), // 20% white
            shadowRadius = 8f,
            shadowOpacity = 0.1f,
            shadowOffsetX = 0f,
            shadowOffsetY = 4f,
            shadowColor = Color.argb(25, 0, 0, 0) // 10% black
        )
        
        val MEDIUM = GlassmorphismConfig(
            style = GlassmorphismStyle.MEDIUM,
            blurRadius = 20f,
            opacity = 0.15f,
            cornerRadius = 16f,
            borderWidth = 1.5f,
            borderColor = Color.argb(76, 255, 255, 255), // 30% white
            shadowRadius = 12f,
            shadowOpacity = 0.15f,
            shadowOffsetX = 0f,
            shadowOffsetY = 6f,
            shadowColor = Color.argb(38, 0, 0, 0) // 15% black
        )
        
        val HEAVY = GlassmorphismConfig(
            style = GlassmorphismStyle.HEAVY,
            blurRadius = 30f,
            opacity = 0.2f,
            cornerRadius = 20f,
            borderWidth = 2f,
            borderColor = Color.argb(102, 255, 255, 255), // 40% white
            shadowRadius = 16f,
            shadowOpacity = 0.2f,
            shadowOffsetX = 0f,
            shadowOffsetY = 8f,
            shadowColor = Color.argb(51, 0, 0, 0) // 20% black
        )
        
        val FROSTED = GlassmorphismConfig(
            style = GlassmorphismStyle.FROSTED,
            blurRadius = 25f,
            opacity = 0.25f,
            cornerRadius = 18f,
            borderWidth = 1.5f,
            borderColor = Color.argb(127, 255, 255, 255), // 50% white
            shadowRadius = 14f,
            shadowOpacity = 0.18f,
            shadowOffsetX = 0f,
            shadowOffsetY = 7f,
            shadowColor = Color.argb(46, 0, 0, 0) // 18% black
        )
        
        val CRYSTAL = GlassmorphismConfig(
            style = GlassmorphismStyle.CRYSTAL,
            blurRadius = 15f,
            opacity = 0.08f,
            cornerRadius = 14f,
            borderWidth = 0.5f,
            borderColor = Color.argb(153, 255, 255, 255), // 60% white
            shadowRadius = 10f,
            shadowOpacity = 0.12f,
            shadowOffsetX = 0f,
            shadowOffsetY = 5f,
            shadowColor = Color.argb(30, 0, 0, 0) // 12% black
        )
        
        val STORM = GlassmorphismConfig(
            style = GlassmorphismStyle.STORM,
            blurRadius = 22f,
            opacity = 0.18f,
            cornerRadius = 16f,
            borderWidth = 1.5f,
            borderColor = Color.argb(76, 245, 158, 11), // Golden accent
            shadowRadius = 13f,
            shadowOpacity = 0.16f,
            shadowOffsetX = 0f,
            shadowOffsetY = 6f,
            shadowColor = Color.argb(41, 10, 17, 40) // Storm sky color
        )
        
        // Storm Sky themed configurations
        val STORM_SKY_LIGHT = GlassmorphismConfig(
            style = GlassmorphismStyle.STORM,
            blurRadius = 15f,
            opacity = 0.12f,
            cornerRadius = 12f,
            borderWidth = 1f,
            borderColor = Color.argb(51, 245, 158, 11), // 20% golden
            shadowRadius = 8f,
            shadowOpacity = 0.1f,
            shadowOffsetX = 0f,
            shadowOffsetY = 4f,
            shadowColor = Color.argb(25, 10, 17, 40) // 10% storm sky
        )
        
        val STORM_SKY_MEDIUM = GlassmorphismConfig(
            style = GlassmorphismStyle.STORM,
            blurRadius = 22f,
            opacity = 0.18f,
            cornerRadius = 16f,
            borderWidth = 1.5f,
            borderColor = Color.argb(76, 245, 158, 11), // 30% golden
            shadowRadius = 13f,
            shadowOpacity = 0.16f,
            shadowOffsetX = 0f,
            shadowOffsetY = 6f,
            shadowColor = Color.argb(41, 10, 17, 40) // 16% storm sky
        )
        
        val STORM_SKY_HEAVY = GlassmorphismConfig(
            style = GlassmorphismStyle.STORM,
            blurRadius = 30f,
            opacity = 0.25f,
            cornerRadius = 20f,
            borderWidth = 2f,
            borderColor = Color.argb(102, 245, 158, 11), // 40% golden
            shadowRadius = 18f,
            shadowOpacity = 0.22f,
            shadowOffsetX = 0f,
            shadowOffsetY = 8f,
            shadowColor = Color.argb(56, 10, 17, 40) // 22% storm sky
        )
    }
}

// MARK: - Glassmorphism Drawable
class GlassmorphismDrawable(
    private val config: GlassmorphismConfig,
    private val context: Context
) : Drawable() {
    
    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val shadowPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    
    init {
        setupPaints()
    }
    
    private fun setupPaints() {
        // Background paint
        paint.color = Color.argb(
            (255 * config.opacity).toInt(),
            255, 255, 255
        )
        paint.style = Paint.Style.FILL
        
        // Border paint
        borderPaint.color = config.borderColor
        borderPaint.style = Paint.Style.STROKE
        borderPaint.strokeWidth = config.borderWidth
        
        // Shadow paint
        shadowPaint.color = config.shadowColor
        shadowPaint.style = Paint.Style.FILL
        shadowPaint.maskFilter = BlurMaskFilter(config.shadowRadius, BlurMaskFilter.Blur.NORMAL)
    }
    
    override fun draw(canvas: Canvas) {
        val bounds = bounds
        val rect = RectF(
            bounds.left.toFloat(),
            bounds.top.toFloat(),
            bounds.right.toFloat(),
            bounds.bottom.toFloat()
        )
        
        // Draw shadow
        canvas.drawRoundRect(
            rect.left + config.shadowOffsetX,
            rect.top + config.shadowOffsetY,
            rect.right + config.shadowOffsetX,
            rect.bottom + config.shadowOffsetY,
            config.cornerRadius,
            config.cornerRadius,
            shadowPaint
        )
        
        // Draw background
        canvas.drawRoundRect(rect, config.cornerRadius, config.cornerRadius, paint)
        
        // Draw border
        canvas.drawRoundRect(rect, config.cornerRadius, config.cornerRadius, borderPaint)
    }
    
    override fun setAlpha(alpha: Int) {
        paint.alpha = alpha
        borderPaint.alpha = alpha
        shadowPaint.alpha = alpha
    }
    
    override fun setColorFilter(colorFilter: ColorFilter?) {
        paint.colorFilter = colorFilter
        borderPaint.colorFilter = colorFilter
        shadowPaint.colorFilter = colorFilter
    }
    
    override fun getOpacity(): Int = PixelFormat.TRANSLUCENT
}

// MARK: - Glassmorphism View
class GlassmorphismView @JvmOverloads constructor(
    context: Context,
    private val config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM
) : FrameLayout(context) {
    
    private val glassmorphismDrawable: GlassmorphismDrawable
    
    init {
        glassmorphismDrawable = GlassmorphismDrawable(config, context)
        background = glassmorphismDrawable
    }
    
    fun updateConfig(newConfig: GlassmorphismConfig) {
        val newDrawable = GlassmorphismDrawable(newConfig, context)
        background = newDrawable
    }
    
    fun animateConfigChange(newConfig: GlassmorphismConfig, duration: Long = 300) {
        // Simple animation - in real implementation, use ValueAnimator
        updateConfig(newConfig)
    }
}

// MARK: - Glassmorphism Button
class GlassmorphismButton @JvmOverloads constructor(
    context: Context,
    private val config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM
) : Button(context) {
    
    private val glassmorphismView: GlassmorphismView
    
    init {
        glassmorphismView = GlassmorphismView(context, config)
        setupButton()
    }
    
    private fun setupButton() {
        // Add glassmorphism view as background
        addView(glassmorphismView, 0)
        glassmorphismView.layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        
        // Button styling
        setTextColor(Color.WHITE)
        textSize = 16f
        typeface = Typeface.DEFAULT_BOLD
        
        // Add touch effects
        setOnTouchListener { _, event ->
            when (event.action) {
                android.view.MotionEvent.ACTION_DOWN -> {
                    scaleX = 0.95f
                    scaleY = 0.95f
                    alpha = 0.8f
                }
                android.view.MotionEvent.ACTION_UP,
                android.view.MotionEvent.ACTION_CANCEL -> {
                    scaleX = 1.0f
                    scaleY = 1.0f
                    alpha = 1.0f
                }
            }
            false
        }
    }
    
    fun updateGlassmorphismConfig(newConfig: GlassmorphismConfig) {
        glassmorphismView.updateConfig(newConfig)
    }
}

// MARK: - Glassmorphism Card
class GlassmorphismCard @JvmOverloads constructor(
    context: Context,
    private val config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM
) : FrameLayout(context) {
    
    private val glassmorphismView: GlassmorphismView
    private val contentView: LinearLayout
    
    init {
        glassmorphismView = GlassmorphismView(context, config)
        contentView = LinearLayout(context)
        setupCard()
    }
    
    private fun setupCard() {
        // Add glassmorphism view as background
        addView(glassmorphismView, 0)
        glassmorphismView.layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        
        // Setup content view
        addView(contentView)
        contentView.layoutParams = LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
        contentView.setPadding(16)
        contentView.orientation = LinearLayout.VERTICAL
    }
    
    fun addContent(view: View) {
        contentView.addView(view)
    }
    
    fun updateGlassmorphismConfig(newConfig: GlassmorphismConfig) {
        glassmorphismView.updateConfig(newConfig)
    }
}

// MARK: - Glassmorphism Utility Functions
object GlassmorphismUtils {
    
    fun createGlassmorphismBackground(view: View, config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM) {
        val glassmorphismView = GlassmorphismView(view.context, config)
        (view as? ViewGroup)?.addView(glassmorphismView, 0)
        
        glassmorphismView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
    }
    
    fun createGradientGlassmorphism(
        view: View,
        colors: IntArray,
        config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM
    ) {
        val gradientDrawable = GradientDrawable(
            GradientDrawable.Orientation.TL_BR,
            colors
        )
        gradientDrawable.cornerRadius = config.cornerRadius
        
        val glassmorphismView = GlassmorphismView(view.context, config)
        glassmorphismView.background = gradientDrawable
        
        (view as? ViewGroup)?.addView(glassmorphismView, 0)
        glassmorphismView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
    }
    
    @RequiresApi(Build.VERSION_CODES.S)
    fun createBlurGlassmorphism(
        view: View,
        config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM
    ) {
        val blurEffect = android.graphics.BlurMaskFilter(config.blurRadius, BlurMaskFilter.Blur.NORMAL)
        val glassmorphismView = GlassmorphismView(view.context, config)
        
        // Apply blur effect
        glassmorphismView.layerType = View.LAYER_TYPE_SOFTWARE
        glassmorphismView.setLayerType(View.LAYER_TYPE_SOFTWARE, null)
        
        (view as? ViewGroup)?.addView(glassmorphismView, 0)
        glassmorphismView.layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
    }
}

// MARK: - Glassmorphism Extension Functions
fun View.applyGlassmorphism(config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM) {
    val glassmorphismView = GlassmorphismView(context, config)
    (parent as? ViewGroup)?.let { parentView ->
        val index = parentView.indexOfChild(this)
        parentView.addView(glassmorphismView, index)
        glassmorphismView.layoutParams = layoutParams
        layoutParams = ViewGroup.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.MATCH_PARENT
        )
    }
}

fun Button.applyGlassmorphismButton(config: GlassmorphismConfig = GlassmorphismConfig.MEDIUM) {
    val glassmorphismButton = GlassmorphismButton(context, config)
    glassmorphismButton.text = text
    glassmorphismButton.setTextColor(textColors)
    glassmorphismButton.textSize = textSize
    glassmorphismButton.typeface = typeface
    
    (parent as? ViewGroup)?.let { parentView ->
        val index = parentView.indexOfChild(this)
        parentView.removeView(this)
        parentView.addView(glassmorphismButton, index)
    }
}

File: mobile/ios/UI/GlassmorphismEffects.swift
//
//  GlassmorphismEffects.swift
//  ALADDIN Mobile Security
//
//  Glassmorphism UI Effects for iOS
//  Modern glass-like visual effects with blur and transparency
//
//  Created by ALADDIN Security Team
//  Date: 2025-01-27
//  Version: 1.0
//

import UIKit

// MARK: - Glassmorphism Effect Types
enum GlassmorphismStyle {
    case light      // Light glass effect
    case medium     // Medium glass effect
    case heavy      // Heavy glass effect
    case frosted    // Frosted glass effect
    case crystal    // Crystal clear effect
    case storm      // Storm sky themed effect
}

// MARK: - Glassmorphism Configuration
struct GlassmorphismConfig {
    let style: GlassmorphismStyle
    let blurRadius: CGFloat
    let opacity: CGFloat
    let cornerRadius: CGFloat
    let borderWidth: CGFloat
    let borderColor: UIColor
    let shadowRadius: CGFloat
    let shadowOpacity: Float
    let shadowOffset: CGSize
    let shadowColor: UIColor
    
    static let light = GlassmorphismConfig(
        style: .light,
        blurRadius: 10,
        opacity: 0.1,
        cornerRadius: 12,
        borderWidth: 1,
        borderColor: UIColor.white.withAlphaComponent(0.2),
        shadowRadius: 8,
        shadowOpacity: 0.1,
        shadowOffset: CGSize(width: 0, height: 4),
        shadowColor: UIColor.black
    )
    
    static let medium = GlassmorphismConfig(
        style: .medium,
        blurRadius: 20,
        opacity: 0.15,
        cornerRadius: 16,
        borderWidth: 1.5,
        borderColor: UIColor.white.withAlphaComponent(0.3),
        shadowRadius: 12,
        shadowOpacity: 0.15,
        shadowOffset: CGSize(width: 0, height: 6),
        shadowColor: UIColor.black
    )
    
    static let heavy = GlassmorphismConfig(
        style: .heavy,
        blurRadius: 30,
        opacity: 0.2,
        cornerRadius: 20,
        borderWidth: 2,
        borderColor: UIColor.white.withAlphaComponent(0.4),
        shadowRadius: 16,
        shadowOpacity: 0.2,
        shadowOffset: CGSize(width: 0, height: 8),
        shadowColor: UIColor.black
    )
    
    static let frosted = GlassmorphismConfig(
        style: .frosted,
        blurRadius: 25,
        opacity: 0.25,
        cornerRadius: 18,
        borderWidth: 1.5,
        borderColor: UIColor.white.withAlphaComponent(0.5),
        shadowRadius: 14,
        shadowOpacity: 0.18,
        shadowOffset: CGSize(width: 0, height: 7),
        shadowColor: UIColor.black
    )
    
    static let crystal = GlassmorphismConfig(
        style: .crystal,
        blurRadius: 15,
        opacity: 0.08,
        cornerRadius: 14,
        borderWidth: 0.5,
        borderColor: UIColor.white.withAlphaComponent(0.6),
        shadowRadius: 10,
        shadowOpacity: 0.12,
        shadowOffset: CGSize(width: 0, height: 5),
        shadowColor: UIColor.black
    )
    
    static let storm = GlassmorphismConfig(
        style: .storm,
        blurRadius: 22,
        opacity: 0.18,
        cornerRadius: 16,
        borderWidth: 1.5,
        borderColor: UIColor(red: 0.96, green: 0.58, blue: 0.04, alpha: 0.3), // Golden accent
        shadowRadius: 13,
        shadowOpacity: 0.16,
        shadowOffset: CGSize(width: 0, height: 6),
        shadowColor: UIColor(red: 0.04, green: 0.07, blue: 0.16, alpha: 0.3) // Storm sky color
    )
}

// MARK: - Glassmorphism View
class GlassmorphismView: UIView {
    
    private var blurEffectView: UIVisualEffectView?
    private var config: GlassmorphismConfig
    
    init(config: GlassmorphismConfig = .medium) {
        self.config = config
        super.init(frame: .zero)
        setupGlassmorphismEffect()
    }
    
    required init?(coder: NSCoder) {
        self.config = .medium
        super.init(coder: coder)
        setupGlassmorphismEffect()
    }
    
    private func setupGlassmorphismEffect() {
        // Create blur effect
        let blurEffect = UIBlurEffect(style: .systemUltraThinMaterial)
        blurEffectView = UIVisualEffectView(effect: blurEffect)
        
        guard let blurEffectView = blurEffectView else { return }
        
        // Configure blur view
        blurEffectView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(blurEffectView)
        
        // Set constraints
        NSLayoutConstraint.activate([
            blurEffectView.topAnchor.constraint(equalTo: topAnchor),
            blurEffectView.leadingAnchor.constraint(equalTo: leadingAnchor),
            blurEffectView.trailingAnchor.constraint(equalTo: trailingAnchor),
            blurEffectView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        // Apply glassmorphism styling
        applyGlassmorphismStyle()
    }
    
    private func applyGlassmorphismStyle() {
        // Background
        backgroundColor = UIColor.clear
        
        // Corner radius
        layer.cornerRadius = config.cornerRadius
        clipsToBounds = true
        
        // Border
        layer.borderWidth = config.borderWidth
        layer.borderColor = config.borderColor.cgColor
        
        // Shadow
        layer.shadowRadius = config.shadowRadius
        layer.shadowOpacity = config.shadowOpacity
        layer.shadowOffset = config.shadowOffset
        layer.shadowColor = config.shadowColor.cgColor
        
        // Blur effect opacity
        blurEffectView?.alpha = config.opacity
    }
    
    func updateConfig(_ newConfig: GlassmorphismConfig) {
        self.config = newConfig
        applyGlassmorphismStyle()
    }
    
    func animateGlassmorphism(to newConfig: GlassmorphismConfig, duration: TimeInterval = 0.3) {
        UIView.animate(withDuration: duration, delay: 0, options: [.curveEaseInOut], animations: {
            self.updateConfig(newConfig)
        })
    }
}

// MARK: - Glassmorphism Button
class GlassmorphismButton: UIButton {
    
    private var config: GlassmorphismConfig
    private var glassmorphismView: GlassmorphismView?
    
    init(config: GlassmorphismConfig = .medium) {
        self.config = config
        super.init(frame: .zero)
        setupGlassmorphismButton()
    }
    
    required init?(coder: NSCoder) {
        self.config = .medium
        super.init(coder: coder)
        setupGlassmorphismButton()
    }
    
    private func setupGlassmorphismButton() {
        // Create glassmorphism view
        glassmorphismView = GlassmorphismView(config: config)
        
        guard let glassmorphismView = glassmorphismView else { return }
        
        // Add as background
        insertSubview(glassmorphismView, at: 0)
        glassmorphismView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            glassmorphismView.topAnchor.constraint(equalTo: topAnchor),
            glassmorphismView.leadingAnchor.constraint(equalTo: leadingAnchor),
            glassmorphismView.trailingAnchor.constraint(equalTo: trailingAnchor),
            glassmorphismView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        // Button styling
        setTitleColor(.white, for: .normal)
        titleLabel?.font = UIFont.systemFont(ofSize: 16, weight: .semibold)
        
        // Add touch effects
        addTarget(self, action: #selector(touchDown), for: .touchDown)
        addTarget(self, action: #selector(touchUp), for: [.touchUpInside, .touchUpOutside, .touchCancel])
    }
    
    @objc private func touchDown() {
        UIView.animate(withDuration: 0.1) {
            self.transform = CGAffineTransform(scaleX: 0.95, y: 0.95)
            self.alpha = 0.8
        }
    }
    
    @objc private func touchUp() {
        UIView.animate(withDuration: 0.1) {
            self.transform = .identity
            self.alpha = 1.0
        }
    }
    
    func updateGlassmorphismConfig(_ newConfig: GlassmorphismConfig) {
        glassmorphismView?.updateConfig(newConfig)
    }
}

// MARK: - Glassmorphism Card
class GlassmorphismCard: UIView {
    
    private var config: GlassmorphismConfig
    private var glassmorphismView: GlassmorphismView?
    private var contentView: UIView = UIView()
    
    init(config: GlassmorphismConfig = .medium) {
        self.config = config
        super.init(frame: .zero)
        setupGlassmorphismCard()
    }
    
    required init?(coder: NSCoder) {
        self.config = .medium
        super.init(coder: coder)
        setupGlassmorphismCard()
    }
    
    private func setupGlassmorphismCard() {
        // Create glassmorphism view
        glassmorphismView = GlassmorphismView(config: config)
        
        guard let glassmorphismView = glassmorphismView else { return }
        
        // Add glassmorphism as background
        insertSubview(glassmorphismView, at: 0)
        glassmorphismView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            glassmorphismView.topAnchor.constraint(equalTo: topAnchor),
            glassmorphismView.leadingAnchor.constraint(equalTo: leadingAnchor),
            glassmorphismView.trailingAnchor.constraint(equalTo: trailingAnchor),
            glassmorphismView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        // Setup content view
        contentView.translatesAutoresizingMaskIntoConstraints = false
        addSubview(contentView)
        
        NSLayoutConstraint.activate([
            contentView.topAnchor.constraint(equalTo: topAnchor, constant: 16),
            contentView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 16),
            contentView.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -16),
            contentView.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -16)
        ])
    }
    
    func addContent(_ view: UIView) {
        contentView.addSubview(view)
        view.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            view.topAnchor.constraint(equalTo: contentView.topAnchor),
            view.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
            view.trailingAnchor.constraint(equalTo: contentView.trailingAnchor),
            view.bottomAnchor.constraint(equalTo: contentView.bottomAnchor)
        ])
    }
    
    func updateGlassmorphismConfig(_ newConfig: GlassmorphismConfig) {
        glassmorphismView?.updateConfig(newConfig)
    }
}

// MARK: - Glassmorphism Navigation Bar
extension UINavigationBar {
    
    func applyGlassmorphism(config: GlassmorphismConfig = .light) {
        // Make navigation bar transparent
        setBackgroundImage(UIImage(), for: .default)
        shadowImage = UIImage()
        isTranslucent = true
        
        // Create glassmorphism effect
        let blurEffect = UIBlurEffect(style: .systemUltraThinMaterial)
        let blurEffectView = UIVisualEffectView(effect: blurEffect)
        blurEffectView.alpha = config.opacity
        blurEffectView.translatesAutoresizingMaskIntoConstraints = false
        
        // Add blur effect to navigation bar
        insertSubview(blurEffectView, at: 0)
        
        NSLayoutConstraint.activate([
            blurEffectView.topAnchor.constraint(equalTo: topAnchor),
            blurEffectView.leadingAnchor.constraint(equalTo: leadingAnchor),
            blurEffectView.trailingAnchor.constraint(equalTo: trailingAnchor),
            blurEffectView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        // Apply styling
        layer.cornerRadius = config.cornerRadius
        layer.borderWidth = config.borderWidth
        layer.borderColor = config.borderColor.cgColor
    }
}

// MARK: - Glassmorphism Tab Bar
extension UITabBar {
    
    func applyGlassmorphism(config: GlassmorphismConfig = .light) {
        // Make tab bar transparent
        backgroundImage = UIImage()
        shadowImage = UIImage()
        isTranslucent = true
        
        // Create glassmorphism effect
        let blurEffect = UIBlurEffect(style: .systemUltraThinMaterial)
        let blurEffectView = UIVisualEffectView(effect: blurEffect)
        blurEffectView.alpha = config.opacity
        blurEffectView.translatesAutoresizingMaskIntoConstraints = false
        
        // Add blur effect to tab bar
        insertSubview(blurEffectView, at: 0)
        
        NSLayoutConstraint.activate([
            blurEffectView.topAnchor.constraint(equalTo: topAnchor),
            blurEffectView.leadingAnchor.constraint(equalTo: leadingAnchor),
            blurEffectView.trailingAnchor.constraint(equalTo: trailingAnchor),
            blurEffectView.bottomAnchor.constraint(equalTo: bottomAnchor)
        ])
        
        // Apply styling
        layer.cornerRadius = config.cornerRadius
        layer.borderWidth = config.borderWidth
        layer.borderColor = config.borderColor.cgColor
    }
}

// MARK: - Glassmorphism Utility Functions
class GlassmorphismUtils {
    
    static func createGlassmorphismBackground(for view: UIView, config: GlassmorphismConfig = .medium) {
        let glassmorphismView = GlassmorphismView(config: config)
        view.insertSubview(glassmorphismView, at: 0)
        
        glassmorphismView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            glassmorphismView.topAnchor.constraint(equalTo: view.topAnchor),
            glassmorphismView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            glassmorphismView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            glassmorphismView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
    }
    
    static func animateGlassmorphismTransition(from view1: UIView, to view2: UIView, duration: TimeInterval = 0.5) {
        UIView.transition(from: view1, to: view2, duration: duration, options: [.transitionCrossDissolve], completion: nil)
    }
    
    static func createGradientGlassmorphism(for view: UIView, colors: [UIColor], config: GlassmorphismConfig = .medium) {
        // Create gradient layer
        let gradientLayer = CAGradientLayer()
        gradientLayer.colors = colors.map { $0.cgColor }
        gradientLayer.locations = [0.0, 1.0]
        gradientLayer.startPoint = CGPoint(x: 0, y: 0)
        gradientLayer.endPoint = CGPoint(x: 1, y: 1)
        
        // Create glassmorphism view
        let glassmorphismView = GlassmorphismView(config: config)
        glassmorphismView.layer.insertSublayer(gradientLayer, at: 0)
        
        view.insertSubview(glassmorphismView, at: 0)
        glassmorphismView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            glassmorphismView.topAnchor.constraint(equalTo: view.topAnchor),
            glassmorphismView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            glassmorphismView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            glassmorphismView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
        
        // Update gradient frame
        DispatchQueue.main.async {
            gradientLayer.frame = glassmorphismView.bounds
        }
    }
}

// MARK: - Storm Sky Glassmorphism Theme
extension GlassmorphismConfig {
    
    static let stormSkyLight = GlassmorphismConfig(
        style: .storm,
        blurRadius: 15,
        opacity: 0.12,
        cornerRadius: 12,
        borderWidth: 1,
        borderColor: UIColor(red: 0.96, green: 0.58, blue: 0.04, alpha: 0.2),
        shadowRadius: 8,
        shadowOpacity: 0.1,
        shadowOffset: CGSize(width: 0, height: 4),
        shadowColor: UIColor(red: 0.04, green: 0.07, blue: 0.16, alpha: 0.2)
    )
    
    static let stormSkyMedium = GlassmorphismConfig(
        style: .storm,
        blurRadius: 22,
        opacity: 0.18,
        cornerRadius: 16,
        borderWidth: 1.5,
        borderColor: UIColor(red: 0.96, green: 0.58, blue: 0.04, alpha: 0.3),
        shadowRadius: 13,
        shadowOpacity: 0.16,
        shadowOffset: CGSize(width: 0, height: 6),
        shadowColor: UIColor(red: 0.04, green: 0.07, blue: 0.16, alpha: 0.3)
    )
    
    static let stormSkyHeavy = GlassmorphismConfig(
        style: .storm,
        blurRadius: 30,
        opacity: 0.25,
        cornerRadius: 20,
        borderWidth: 2,
        borderColor: UIColor(red: 0.96, green: 0.58, blue: 0.04, alpha: 0.4),
        shadowRadius: 18,
        shadowOpacity: 0.22,
        shadowOffset: CGSize(width: 0, height: 8),
        shadowColor: UIColor(red: 0.04, green: 0.07, blue: 0.16, alpha: 0.4)
    )
}

